File: src/main/java/com/kgaft/VulkanLib/RenderPipeline/GraphicsRenderPipeline.java
package com.kgaft.VulkanLib.RenderPipeline;

import com.kgaft.VulkanLib.Device.Image.Image;
import com.kgaft.VulkanLib.Device.LogicalDevice.LogicalDevice;
import com.kgaft.VulkanLib.Device.SwapChain;
import com.kgaft.VulkanLib.Pipelines.GraphicsPipeline.Configuration.GraphicsPipelineBuilder;
import com.kgaft.VulkanLib.Pipelines.GraphicsPipeline.GraphicsPipeline;
import com.kgaft.VulkanLib.Shader.Shader;
import com.kgaft.VulkanLib.Utils.LwjglObject;
import com.kgaft.VulkanLib.Utils.VkErrorException;
import org.lwjgl.vulkan.*;

import java.lang.instrument.IllegalClassFormatException;
import java.util.ArrayList;
import java.util.List;

import static org.lwjgl.vulkan.KHRSwapchain.VK_IMAGE_LAYOUT_PRESENT_SRC_KHR;
import static org.lwjgl.vulkan.VK13.*;
import static org.lwjgl.vulkan.KHRDynamicRendering.*;
public class GraphicsRenderPipeline {

    private static RenderImagePool imagesPool = null;

    private static void checkImagesPool(LogicalDevice device){
        if(imagesPool==null){
            imagesPool = new RenderImagePool(device);
        }
    }

    private LogicalDevice device;
    private SwapChain swapChain;
    private List<Image> baseRenderImages = new ArrayList<>();
    private List<Image> baseDepthImages = new ArrayList<>();
    private GraphicsPipeline graphicsPipeline;

    private LwjglObject<VkRenderingAttachmentInfoKHR.Buffer> colorInfos;
    private LwjglObject<VkRenderingAttachmentInfoKHR> depthInfo;
    private LwjglObject<VkClearValue> colorClear = new LwjglObject<>(VkClearValue.class);
    private LwjglObject<VkClearValue> depthClear = new LwjglObject<>(VkClearValue.class);
    private LwjglObject<VkRenderingInfoKHR> renderingInfoKhr = new LwjglObject<>(VkRenderingInfoKHR.class);

    private int imagePerStepAmount;

    private LwjglObject<VkExtent2D> renderArea;

    private LwjglObject<VkImageMemoryBarrier.Buffer> barriers;

    private boolean firstRender = true;
    private boolean forSwapChain = false;
    private  boolean d = false;

    public GraphicsRenderPipeline(LogicalDevice device, GraphicsPipelineBuilder builder,
                                  Shader shader,
                                  LwjglObject<VkExtent2D> renderArea,
                                  int maxFramesInFlight, int attachmentsPerStepAmount) throws IllegalClassFormatException, VkErrorException {
        checkImagesPool(device);
        boolean populated =false;
        this.imagePerStepAmount = attachmentsPerStepAmount;


        for (int i = 0; i < maxFramesInFlight; ++i) {
            for (int ii = 0; ii < attachmentsPerStepAmount; ++ii) {
                Image colorImage = imagesPool.acquireColorImage(renderArea.get().width(), renderArea.get().height());
                baseRenderImages.add(colorImage);
                if (!populated)
                    builder.addColorAttachmentInfo(colorImage.getImageInfo().get().format());
            }
            Image depthImage = imagesPool.acquireDepthImage(renderArea.get().width(), renderArea.get().height());
            baseDepthImages.add(depthImage);
            builder.setDepthAttachmentInfo(depthImage.getImageInfo().get().format());
            populated = true;
            depthImage.transitionImageLayout(device, VK_IMAGE_LAYOUT_UNDEFINED, VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL, VK_IMAGE_ASPECT_DEPTH_BIT);
        }
        graphicsPipeline = new GraphicsPipeline(builder, attachmentsPerStepAmount, device, shader, renderArea.get().width(), renderArea.get().height());
        this.renderArea = renderArea;

        this.depthClear.get().depthStencil().depth(1);
        this.depthClear.get().depthStencil().stencil(0);
        this.colorClear.get().color().float32(0, 0);
        this.colorClear.get().color().float32(1, 0);
        this.colorClear.get().color().float32(2, 0);
        this.colorClear.get().color().float32(3, 1);
        createImagesAndRenderingInfos(attachmentsPerStepAmount);
    }

    public GraphicsRenderPipeline(LogicalDevice device,
                           SwapChain swapChain,
                                  GraphicsPipelineBuilder pBuilder,
                           Shader shader,
                                  LwjglObject<VkExtent2D> renderArea,
                           int maxFramesInFlight) throws IllegalClassFormatException, VkErrorException {
        this.forSwapChain = true;
        this.swapChain = swapChain;
        this.imagePerStepAmount = 1;
        this.renderArea = renderArea;
        checkImagesPool(device);

        pBuilder.addColorAttachmentInfo(SwapChain.getFormat().format());

        for (int i = 0; i < maxFramesInFlight; ++i) {
            Image depthImage = imagesPool.acquireDepthImage(renderArea.get().width(), renderArea.get().height());
            pBuilder.setDepthAttachmentInfo(depthImage.getImageInfo().get().format());
            baseDepthImages.add(depthImage);

        }
        graphicsPipeline = new GraphicsPipeline(pBuilder, 1, device,
                shader, renderArea.get().width(),
                renderArea.get().height());
        this.depthClear.get().depthStencil().depth(1);
        this.depthClear.get().depthStencil().stencil(0);
        this.colorClear.get().color().float32(0, 0);
        this.colorClear.get().color().float32(1, 0);
        this.colorClear.get().color().float32(2, 0);
        this.colorClear.get().color().float32(3, 1);
        createImagesAndRenderingInfos(1);
    }


    public void begin(VkCommandBuffer cmd, int currentImage) {
        prepareBarriersBeforeRendering(currentImage);
        bindBarriers(cmd);
        colorInfos.get().rewind();
        depthInfo.get().imageView(baseDepthImages.get(currentImage).getImageViews().get(0).getImageView());
        if (forSwapChain) {
            colorInfos.get().get(0).imageView(swapChain.getSwapchainImageViews().get(currentImage).getImageView());
        } else {
            for (int i = currentImage * imagePerStepAmount; i < (currentImage + 1) * imagePerStepAmount; i++) {
                int infoIndex = i - currentImage * imagePerStepAmount;
                colorInfos.get().get(infoIndex).imageView(baseRenderImages.get(i).getImageViews().get(0).getImageView());
            }
        }
        vkCmdBeginRendering(cmd, renderingInfoKhr.get());
        vkCmdBindPipeline(cmd, VK_PIPELINE_BIND_POINT_GRAPHICS, graphicsPipeline.getGraphicsPipeline());
    }

    public void endRender(VkCommandBuffer cmd, int currentImage) {
        vkCmdEndRendering(cmd);
        prepareBarriersAfterRendering(currentImage);
        bindBarriers(cmd);
    }

    public void resize(int width, int height) throws IllegalClassFormatException, VkErrorException {
        d = false;
        firstRender = true;
        renderArea.get().width(width);
        renderArea.get().height(height);
        renderingInfoKhr.get().renderArea().extent().width(width);
        renderingInfoKhr.get().renderArea().extent().height(height);
        if (!forSwapChain) {
            baseRenderImages.forEach(element-> {
                try {
                    element.resize(width, height);
                } catch (VkErrorException e) {
                    throw new RuntimeException(e);
                }
            });

        }
        baseDepthImages.forEach(element-> {
            try {
                element.resize(width,height);
            } catch (VkErrorException e) {
                throw new RuntimeException(e);
            }
        });

        graphicsPipeline.recreate(width, height);

    }

    private void bindBarriers(VkCommandBuffer cmd) {
        vkCmdPipelineBarrier(cmd, VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT, VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT,0,null, null, barriers.get());
    }


    private void createImagesAndRenderingInfos(int imagePerStepAmount) throws IllegalClassFormatException {
        colorInfos = new LwjglObject<>(VkRenderingAttachmentInfoKHR.class, VkRenderingAttachmentInfoKHR.Buffer.class, imagePerStepAmount);
        colorInfos.get().forEach(item->{
            item.sType$Default();
            item.imageLayout(VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL);
            item.loadOp(VK_ATTACHMENT_LOAD_OP_CLEAR);
            item.storeOp(VK_ATTACHMENT_STORE_OP_STORE);
            item.clearValue(colorClear.get());
        });
        colorInfos.get().rewind();
        depthInfo.get().sType$Default();
        depthInfo.get().imageLayout(VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL);
        depthInfo.get().loadOp(VK_ATTACHMENT_LOAD_OP_CLEAR);
        depthInfo.get().storeOp(VK_ATTACHMENT_STORE_OP_STORE);
        depthInfo.get().clearValue(depthClear.get());

        renderingInfoKhr.get().renderArea().extent(renderArea.get());
        renderingInfoKhr.get().layerCount(1);

        renderingInfoKhr.get().pColorAttachments(colorInfos.get());
        renderingInfoKhr.get().pDepthAttachment(depthInfo.get());

        barriers = new LwjglObject<>(VkImageMemoryBarrier.class, VkImageMemoryBarrier.Buffer.class, imagePerStepAmount);

    }
    private void prepareBarriersBeforeRendering(int currentImage) {
        if (firstRender) firstRender = !(d && currentImage == 0);
        for (int i = currentImage * imagePerStepAmount; i < (currentImage + 1) * imagePerStepAmount; i++) {
            int barIndex = i - currentImage * imagePerStepAmount;
            barriers.get().get(barIndex).sType$Default();
            barriers.get().get(barIndex).srcAccessMask(VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT);
            barriers.get().get(barIndex).oldLayout(firstRender ? (forSwapChain ? VK_IMAGE_LAYOUT_UNDEFINED
                    : VK_IMAGE_LAYOUT_GENERAL) : (forSwapChain
                    ? VK_IMAGE_LAYOUT_PRESENT_SRC_KHR : VK_IMAGE_LAYOUT_GENERAL));
            barriers.get().get(barIndex).newLayout(VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL);
            barriers.get().get(barIndex).image(forSwapChain ? swapChain.getSwapchainImages().get(i).getBase()
                                                    : baseRenderImages.get(i).getBase());

            barriers.get().get(barIndex).subresourceRange().aspectMask(VK_IMAGE_ASPECT_COLOR_BIT);
            barriers.get().get(barIndex).subresourceRange().levelCount(1);
            barriers.get().get(barIndex).subresourceRange().layerCount(1);
        }

        d = true;
    }
    private void prepareBarriersAfterRendering(int currentImage) {
        if (firstRender) firstRender = !(d && currentImage == 0);
        for (int i = currentImage * imagePerStepAmount; i < (currentImage + 1) * imagePerStepAmount; i++) {
            int barIndex = i - currentImage * imagePerStepAmount;
            barriers.get().get(barIndex).sType$Default();
            barriers.get().get(barIndex).srcAccessMask(VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT);
            barriers.get().get(barIndex).oldLayout(VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL);
            barriers.get().get(barIndex).newLayout(forSwapChain ? VK_IMAGE_LAYOUT_PRESENT_SRC_KHR : VK_IMAGE_LAYOUT_GENERAL);
            barriers.get().get(barIndex).image(forSwapChain ? swapChain.getSwapchainImages().get(i).getBase()
                    : baseRenderImages.get(i).getBase());

            barriers.get().get(barIndex).subresourceRange().aspectMask(VK_IMAGE_ASPECT_COLOR_BIT);
            barriers.get().get(barIndex).subresourceRange().levelCount(1);
            barriers.get().get(barIndex).subresourceRange().layerCount(1);
        }

        d = true;
    }

}

File: src/cxx/VulkanLib/RenderPipeline/GraphicsRenderPipeline.cpp
//
// Created by kgaft on 10/14/24.
//
#include "GraphicsRenderPipeline.hpp"

RenderPipelineBuilder::RenderPipelineBuilder() {
    pGraphicsPipelineBuilder = new GraphicsPipelineBuilder;
}

void RenderPipelineBuilder::addVertexInput(VertexInput input) {
    pGraphicsPipelineBuilder->addVertexInput(input);
}

void RenderPipelineBuilder::addPushConstantInfo(PushConstantInfo info) {
    pGraphicsPipelineBuilder->addPushConstantInfo(info);
}

void RenderPipelineBuilder::addUniformBuffer(UniformBufferInfo info) {
    pGraphicsPipelineBuilder->addUniformBuffer(info);
}

void RenderPipelineBuilder::addSamplerInfo(SamplerInfo info) {
    pGraphicsPipelineBuilder->addSamplerInfo(info);
}

void RenderPipelineBuilder::addStorageBufferInfo(StorageBufferInfo info) {
    pGraphicsPipelineBuilder->addStorageBufferInfo(info);
}

void RenderPipelineBuilder::setAttachmentsPerStepAmount(uint32_t attachmentsPerStepAmount) {
    RenderPipelineBuilder::attachmentsPerStepAmount = attachmentsPerStepAmount;
}

void RenderPipelineBuilder::clear() {
    delete pGraphicsPipelineBuilder;
    RenderPipelineBuilder::pGraphicsPipelineBuilder = new GraphicsPipelineBuilder;
    attachmentsPerStepAmount = 0;
}

RenderPipelineBuilder::~RenderPipelineBuilder() {
    delete pGraphicsPipelineBuilder;
}

void GraphicsRenderPipeline::checkPool(std::shared_ptr<LogicalDevice> device) {
    if (!initializedRenderImagePool) {
        renderImages = std::make_shared<RenderImagePool>(device);
        initializedRenderImagePool = true;
    }
}

void GraphicsRenderPipeline::releaseRenderImagePool() {
    renderImages->destroy();
    initializedRenderImagePool = false;
}

std::shared_ptr<RenderImagePool> GraphicsRenderPipeline::getRenderImagePool(std::shared_ptr<LogicalDevice> device) {
    checkPool(device);
    return renderImages;
}

GraphicsRenderPipeline::GraphicsRenderPipeline(Instance &instance, std::shared_ptr<LogicalDevice> device,
    RenderPipelineBuilder *pBuilder, Shader *shader, vk::Extent2D renderArea, uint32_t maxFramesInFlight): instance(instance), imagePerStepAmount(pBuilder->attachmentsPerStepAmount), renderArea(renderArea) {
    checkPool(device);
    bool populated = false;
    for (uint32_t i = 0; i < maxFramesInFlight; ++i) {
        for (uint32_t ii = 0; ii < pBuilder->attachmentsPerStepAmount; ++ii) {
            auto colorImage = renderImages->acquireColorRenderImage(renderArea.width, renderArea.height);
            baseRenderImages.push_back(colorImage);
            if (!populated)
                pBuilder->pGraphicsPipelineBuilder->addColorAttachmentInfo(colorImage->getImageInfo().format);
        }
        auto depthImage = renderImages->acquireDepthImage(renderArea.width, renderArea.height);
        baseDepthImages.push_back(depthImage);
        pBuilder->pGraphicsPipelineBuilder->setDepthAttachmentInfo(depthImage->getImageInfo().format);
        populated = true;
        depthImage->transitionImageLayout(device, vk::ImageLayout::eUndefined,
                                          vk::ImageLayout::eDepthStencilAttachmentOptimal,
                                          vk::ImageAspectFlagBits::eDepth);
    }
    graphicsPipeline = std::make_shared<GraphicsPipeline>(*device, shader, pBuilder->pGraphicsPipelineBuilder,
                                                          pBuilder->attachmentsPerStepAmount, renderArea.width,
                                                          renderArea.height);
    depthClear.depthStencil.depth = 1.0f;
    depthClear.depthStencil.stencil = 0;
    colorClear.color = {0.0f, 0.0f, 0.0f, 1.0f};
    viewport.maxDepth = 1.0f;
    viewport.width = renderArea.width;
    viewport.height = renderArea.height;
    scissor.extent.width = renderArea.width;
    scissor.extent.height = renderArea.height;
    createImagesAndRenderingInfos(pBuilder->attachmentsPerStepAmount);

}

GraphicsRenderPipeline::GraphicsRenderPipeline(Instance &instance, std::shared_ptr<LogicalDevice> device,
    std::shared_ptr<SwapChain> swapChain, RenderPipelineBuilder *pBuilder, Shader *shader, vk::Extent2D renderArea,
    uint32_t maxFramesInFlight):
    swapChain(swapChain), instance(instance), imagePerStepAmount(1), renderArea(renderArea),
    forSwapChain(true) {
    checkPool(device);

    pBuilder->pGraphicsPipelineBuilder->addColorAttachmentInfo(swapChain->getFormat().format);
    for (uint32_t i = 0; i < maxFramesInFlight; ++i) {
        auto depthImage = renderImages->acquireDepthImage(renderArea.width, renderArea.height);
        pBuilder->pGraphicsPipelineBuilder->setDepthAttachmentInfo(depthImage->getImageInfo().format);
        baseDepthImages.push_back(depthImage);

    }
    graphicsPipeline = std::make_shared<GraphicsPipeline>(*device, shader, pBuilder->pGraphicsPipelineBuilder,
                                                          1, renderArea.width,
                                                          renderArea.height);
    depthClear.depthStencil.depth = 1.0f;
    depthClear.depthStencil.stencil = 0;
    colorClear.color = {0.0f, 0.0f, 0.0f, 1.0f};
    viewport.width = renderArea.width;
    viewport.height = renderArea.height;
    scissor.extent.width = renderArea.width;
    scissor.extent.height = renderArea.height;
    createImagesAndRenderingInfos(1);
}

void GraphicsRenderPipeline::begin(vk::CommandBuffer cmd, uint32_t currentImage) {
    prepareBarriersBeforeRendering(currentImage);
    bindBarriers(cmd);
    depthInfo.imageView = baseDepthImages[currentImage]->getImageViews()[0]->getBase();
    if (forSwapChain) {
        colorInfos[0].imageView = swapChain->getSwapchainImageViews()[currentImage]->getBase();
    } else {
        for (uint32_t i = currentImage * imagePerStepAmount; i < (currentImage + 1) * imagePerStepAmount; i++) {
            uint32_t infoIndex = i - currentImage * imagePerStepAmount;
            colorInfos[infoIndex].imageView = baseRenderImages[i]->getImageViews()[0]->getBase();
        }
    }
    cmd.beginRenderingKHR(renderingInfoKhr, instance.getDynamicLoader());
    cmd.setViewport(0, 1, &viewport, instance.getDynamicLoader());
    cmd.setScissor(0, 1, &scissor, instance.getDynamicLoader());
    cmd.bindPipeline(vk::PipelineBindPoint::eGraphics, graphicsPipeline->getGraphicsPipeline());

}

void GraphicsRenderPipeline::endRender(vk::CommandBuffer cmd, uint32_t currentImage) {
    cmd.endRenderingKHR(instance.getDynamicLoader());
    prepareBarriersAfterRendering(currentImage);
    bindBarriers(cmd);
}

void GraphicsRenderPipeline::resize(uint32_t width, uint32_t height) {
    d = false;
    firstRender = true;
    renderArea.width = width;
    renderArea.height = height;
    renderingInfoKhr.renderArea.extent.width = width;
    renderingInfoKhr.renderArea.extent.height = height;
    scissor.extent.width = width;
    scissor.extent.height = height;
    viewport.width = renderArea.width;
    viewport.height = renderArea.height;
    if (!forSwapChain)  {
        for (auto &item: baseRenderImages) {
            item->resize(width, height);
        }
    }
    for (auto &item: baseDepthImages) {
        item->resize(width, height);
    }

}

const std::vector<std::shared_ptr<Image>> & GraphicsRenderPipeline::getBaseRenderImages() const {
    return baseRenderImages;
}

const std::vector<std::shared_ptr<Image>> & GraphicsRenderPipeline::getBaseDepthImages() const {
    return baseDepthImages;
}

const std::shared_ptr<GraphicsPipeline> & GraphicsRenderPipeline::getGraphicsPipeline() const {
    return graphicsPipeline;
}

void GraphicsRenderPipeline::bindBarriers(vk::CommandBuffer cmd) {
    for (const auto &item: barriers) {
        cmd.pipelineBarrier(vk::PipelineStageFlagBits::eColorAttachmentOutput,
                            vk::PipelineStageFlagBits::eTopOfPipe,
                            vk::DependencyFlags(), 0, 0,
                            item
        );
    }
}

void GraphicsRenderPipeline::createImagesAndRenderingInfos(uint32_t imagePerStepAmount) {
    colorInfos.resize(imagePerStepAmount);
    for (auto &item: colorInfos) {
        item.sType = vk::StructureType::eRenderingAttachmentInfoKHR;
        item.imageLayout = vk::ImageLayout::eColorAttachmentOptimal;
        item.loadOp = vk::AttachmentLoadOp::eClear;
        item.storeOp = vk::AttachmentStoreOp::eStore;
        item.clearValue = colorClear;
    }
    depthInfo.sType = vk::StructureType::eRenderingAttachmentInfoKHR;
    depthInfo.imageLayout = vk::ImageLayout::eDepthStencilAttachmentOptimal;
    depthInfo.loadOp = vk::AttachmentLoadOp::eClear;
    depthInfo.storeOp = vk::AttachmentStoreOp::eStore;
    depthInfo.clearValue = depthClear;

    renderingInfoKhr.renderArea = {0, 0, renderArea.width, renderArea.height};
    renderingInfoKhr.layerCount = 1;
    renderingInfoKhr.colorAttachmentCount = colorInfos.size();
    renderingInfoKhr.pColorAttachments = colorInfos.data();
    renderingInfoKhr.pDepthAttachment = &depthInfo;

    barriers.resize(imagePerStepAmount);

}

void GraphicsRenderPipeline::prepareBarriersBeforeRendering(uint32_t currentImage) {
    if (firstRender) firstRender = !(d && currentImage == 0);
    for (uint32_t i = currentImage * imagePerStepAmount; i < (currentImage + 1) * imagePerStepAmount; i++) {
        uint32_t barIndex = i - currentImage * imagePerStepAmount;
        barriers[barIndex].sType = vk::StructureType::eImageMemoryBarrier;
        barriers[barIndex].srcAccessMask = vk::AccessFlagBits::eColorAttachmentWrite;
        barriers[barIndex].oldLayout = firstRender ? (forSwapChain ? vk::ImageLayout::eUndefined
                                                          : vk::ImageLayout::eGeneral) : (forSwapChain
                                               ? vk::ImageLayout::ePresentSrcKHR
                                               : vk::ImageLayout::eGeneral);
        barriers[barIndex].newLayout = vk::ImageLayout::eColorAttachmentOptimal;
        barriers[barIndex].image = forSwapChain ? swapChain->getSwapchainImages()[i]->getBase()
                                       : baseRenderImages[i]->getBase();

        barriers[barIndex].subresourceRange = {
            .aspectMask = VK_IMAGE_ASPECT_COLOR_BIT,
            .baseMipLevel = 0,
            .levelCount = 1,
            .baseArrayLayer = 0,
            .layerCount = 1,
        };
    }

    d = true;
}

void GraphicsRenderPipeline::prepareBarriersAfterRendering(uint32_t currentImage) {
    for (uint32_t i = currentImage * imagePerStepAmount; i < (currentImage + 1) * imagePerStepAmount; i++) {
        uint32_t barIndex = i - currentImage * imagePerStepAmount;
        barriers[barIndex].srcAccessMask = vk::AccessFlagBits::eColorAttachmentWrite;
        barriers[barIndex].oldLayout = vk::ImageLayout::eColorAttachmentOptimal;
        barriers[barIndex].newLayout = forSwapChain ? vk::ImageLayout::ePresentSrcKHR : vk::ImageLayout::eGeneral;
        barriers[barIndex].image = forSwapChain ? swapChain->getSwapchainImages()[i]->getBase()
                                       : baseRenderImages[i]->getBase();

        barriers[barIndex].subresourceRange = {
            .aspectMask = VK_IMAGE_ASPECT_COLOR_BIT,
            .baseMipLevel = 0,
            .levelCount = 1,
            .baseArrayLayer = 0,
            .layerCount = 1,
        };

    }
}

void GraphicsRenderPipeline::destroy() {
    destroyed = true;
    if (!forSwapChain) {
        for (auto &item: baseRenderImages) {
            renderImages->releaseColorImage(item);
        }
    }
    for (auto &item: baseDepthImages) {
        renderImages->releaseDepthImage(item);
    }
    graphicsPipeline->destroy();
}
